File: Semestre_2/Vision_Computadora/Entregas/Matlab/T2_9_Operadores_Booleanos.m
close all;
clc;

[file,path] = uigetfile({'*.jpg';'*.png';'*.jpeg';'*.bmp';'*.gif'},'File Selector');
ruta = fullfile(path,file);

A = imread(ruta);
A = im2double(A);
A_color = A; 
A = rgb2gray(A);
A_Gris = A;
A = imbinarize(A);

[file2,path2] = uigetfile({'*.jpg';'*.png';'*.jpeg';'*.bmp';'*.gif'},'File Selector');
ruta2 = fullfile(path2,file2);

B = imread(ruta2);
B = im2double(B);
B_color = B;
B = rgb2gray(B);
B_Gris = B;
B = imbinarize(B);

%% --- COLOR ---
[bx, by, ~] = size(B_color);

% Asegurar que ambas sean RGB (3 canales)
if size(A_color,3) == 1
    A_color = cat(3, A_color, A_color, A_color);
end
if size(B_color,3) == 1
    B_color = cat(3, B_color, B_color, B_color);
end

A_color = imresize(A_color, [bx by]);

mask = (A_color > 0) & (B_color > 0);

C_color = A_color;
C_color(~mask) = 0;

figure,
subplot(1,3,1); imshow(A_color); title('Imagen A (color)');
subplot(1,3,2); imshow(B_color); title('Imagen B (color)');
subplot(1,3,3); imshow(C_color); title('A & B (color)');


%% --- GRIS ---
[bx,by] = size(B_Gris);
A_Gris = imresize(A_Gris,[bx by]);
C_Gris = A_Gris & B_Gris;

figure;
subplot(1,3,1); imshow(A_Gris); title('Imagen A (gris)');
subplot(1,3,2); imshow(B_Gris); title('Imagen B (gris)');
subplot(1,3,3); imshow(C_Gris); title('A & B (gris)');

%% --- BOOLEANO ---
[bx,by] = size(B);
A = imresize(A,[bx by]);
C = A & B;

figure;
subplot(1,3,1); imshow(A); title('Imagen A (binaria)');
subplot(1,3,2); imshow(B); title('Imagen B (binaria)');
subplot(1,3,3); imshow(C); title('A & B (binaria)');

File: Semestre_2/Vision_Computadora/Entregas/Python/T2_9_Operadores_Booleanos.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from skimage.color import rgb2gray
from skimage.filters import threshold_otsu
from tkinter import Tk, filedialog
from skimage.exposure import equalize_hist, equalize_adapthist
from skimage.transform import resize

Tk().withdraw()
file_path1 = filedialog.askopenfilename(
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)

A = imread(file_path1).astype(float) / 255.0
A_color = A.copy()
A_Gris = rgb2gray(A)
A_bin = A_Gris > threshold_otsu(A_Gris)

file_path2 = filedialog.askopenfilename(
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)

B = imread(file_path2).astype(float) / 255.0
B_color = B.copy()
B_Gris = rgb2gray(B)
B_bin = B_Gris > threshold_otsu(B_Gris)

bx, by, _ = B_color.shape

if A_color.ndim == 2:
    A_color = np.dstack([A_color, A_color, A_color])
if B_color.ndim == 2:
    B_color = np.dstack([B_color, B_color, B_color])

A_color = resize(A_color, (bx, by), preserve_range=True)

mask = (A_color > 0) & (B_color > 0)

C_color = A_color.copy()
C_color[~mask] = 0

plt.figure()
plt.subplot(1,3,1); plt.imshow(A_color); plt.title('Imagen A (color)'); plt.axis('off')
plt.subplot(1,3,2); plt.imshow(B_color); plt.title('Imagen B (color)'); plt.axis('off')
plt.subplot(1,3,3); plt.imshow(C_color); plt.title('A & B (color)'); plt.axis('off')

bx, by = B_Gris.shape
A_Gris_res = resize(A_Gris, (bx, by), preserve_range=True)
C_Gris = (A_Gris_res > 0) & (B_Gris > 0)

plt.figure()
plt.subplot(1,3,1); plt.imshow(A_Gris_res, cmap='gray'); plt.title('Imagen A (gris)'); plt.axis('off')
plt.subplot(1,3,2); plt.imshow(B_Gris, cmap='gray'); plt.title('Imagen B (gris)'); plt.axis('off')
plt.subplot(1,3,3); plt.imshow(C_Gris, cmap='gray'); plt.title('A & B (gris)'); plt.axis('off')

bx, by = B_bin.shape
A_bin_res = resize(A_bin.astype(float), (bx, by), preserve_range=True) > 0.5
C_bin = A_bin_res & B_bin

plt.figure()
plt.subplot(1,3,1); plt.imshow(A_bin_res, cmap='gray'); plt.title('Imagen A (binaria)'); plt.axis('off')
plt.subplot(1,3,2); plt.imshow(B_bin, cmap='gray'); plt.title('Imagen B (binaria)'); plt.axis('off')
plt.subplot(1,3,3); plt.imshow(C_bin, cmap='gray'); plt.title('A & B (binaria)'); plt.axis('off')
plt.show()
